src/condor_shadow.V6.1/NTreceivers.cpp

```diff
@@ -2314,6 +2316,124 @@ case CONDOR_getdir:
 		return put_x509_rc;
 	}
 
+	case CONDOR_get_docker_creds:
+	{
+#ifdef WIN32
+		ASSERT("docker credentials not supported on WINDOWS");
+#else
+		// The starter will send us a queryAd, currently empty
+		ClassAd queryAd;
+		result = getClassAd(syscall_sock, queryAd);
+		ASSERT(result)
+		result = syscall_sock->end_of_message();
+		ASSERT(result);
+		dprintf(D_SYSCALLS, "\t docker_creds_query ad has %d entries\n", queryAd.size());
+
+
+		ClassAd *job_ad;
+		pseudo_get_job_ad( job_ad );
+		bool wants_creds = false;
+
+		// Double check that the job wanted the credentials, and that
+		// no one is asking for what they should not be asking for.
+		job_ad->LookupBool(ATTR_DOCKER_SEND_CREDENTIALS, wants_creds);
+		if (!wants_creds) {
+			ClassAd authsAd; // The ad to return
+			authsAd.Assign("HTCondorError", "Job did not request docker credentials to be sent");
+			syscall_sock->encode();
+			result = putClassAd(syscall_sock, authsAd);
+			ASSERT( result );
+			result = syscall_sock->end_of_message();
+			ON_ERROR_RETURN( result );
+		}
+
+		// And we will send back a docker config.json file, encoded as a classad
+		// with only the auths included.
+		//
+		// Unless there is an error, in which case, we will add as
+		// "HTCondorError" field with a description of the problem.
+
+		int exit_status = 0;
+		const int timeout = 120;
+		ArgList args;
+
+		char *pat_cstr = param("DOCKER_PAT_PRODUCER");
+		if (!pat_cstr) {
+			EXCEPT("DOCKER_PAT_PRODUCER not defined in config file");
+		}
+		args.AppendArg(pat_cstr);
+		free(pat_cstr);
+		Env command_env;
+		Env shadow_env;
+		shadow_env.Import();
+		std::string path;
+		shadow_env.GetEnv("PATH", path);
+		command_env.SetEnv("PATH", path);
+#if !defined(WIN32)
+		struct passwd *pw = getpwuid( get_user_uid() );
+		if ( pw && pw->pw_dir ) {
+			command_env.SetEnv( "HOME", pw->pw_dir );
+		} else {
+			dprintf( D_ALWAYS, "Failed to find user's home directory to set HOME for docker pat producer\n" );
+		}
+#endif
+
+		// Run the command with a 120 second timeout
+		char *out = run_command(timeout, args, 0 /*options*/, &command_env, &exit_status); 
+
+		std::string creds_json = out;
+		free(out);
+
+		classad::ClassAdJsonParser cajp;
+		ClassAd *pat_ad = cajp.ParseClassAd(creds_json);
```

Comment:
Better to use the ParseClassAd() variant that takes a ClassAd as a reference (less explicit memory management).